Expose a helper to deallocate thread stack

jrvidal opened the pull request:
This is not ready to merge, as there are a couple of TODOs yet, but I wanted to validate the rough idea.

We use the global allocator to request space for the thread stack instead of requesting a page directly. In this way, the allocator takes care of synchronization issues when calling `free()`.

Note that we're relying on the global allocator to be aware of multi-threading, which is the case for the one shipped by `std` in the `"atomics"` target. I think this is not specially problematic, since the current stack allocation shenanigans already assume certain things about the current allocator: namely, that the heap is not entirely under its control (since we do request pages outside of it!)

This doesn't run any destructors or dispose of TLS storage, which seems like a significantly more difficult problem. But I suppose that this patch would not preclude a future, more complete solution.

Fixes #2764

alexcrichton commented:
Hm I don't think that this strategy works when spawning threads concurrently because I think they'll all use the same stack? When a thread starts its default stack pointer is the module-default stack pointer which is presumably reserved for the "main thread", so to call `malloc` this would have to already switch to a separate stack or otherwise be synchronized I think?

jrvidal commented:
:facepalm: What a silly mistake. OK, I've added some extra space for a shared, "scratch-space" stack that can threads can grab temporarily just for calls to `malloc()`/`free()`.

One thing that I've realized is this all sort of assumes that if an agent that cannot block (`[[CanBlock]] == false`, e.g. the main thread) instantiates the module, it has to be the first "thread". Otherwise there's risk of calling `atomic.wait` and getting an error. This is sort of a "breaking" change and I wonder all these new initialization should be opt-in under an environment variable (a la `WASM_BINDGEN_THREADS`).

jrvidal commented:
OK, I've added an environment variable to opt-in, and also exposed the intrinsic properly after gc passes. I'm not totally sure I'm re-injecting `__wbdingen_thread_destroy` in the right place though.

alexcrichton commented:
Ok this looks pretty reasonable to me! I'm a bit sad about the hard-coded lock here but I think it's somewhat inevitable as well. As this gets a little unwieldy over time would you be up for adding some tests along the lines of the externref-xform? Basically I'd like to make sure this runs on CI and ideally see the `*.wat` output of the transformations to ensure that it's correrct.

Additionally I still think that something should probably be done with TLS. Wasm-bindgen is injecting TLS inititalization glue (and the malloc) so it's ostensibly responsible for cleaning that up. I think that the `__wbindgen_thread_destroy` can basically just start, on the thread's normal stack, to call free for the tls glue allocation. I think you might need to add a "convention" for finding the TLS glue base global, but it's basically the same as finding the stack pointer.

A major missing piece with TLS is destruction of each individual value. This needs coordination with the standard library, however, and would involve [fixing this FIXME](https://github.com/rust-lang/rust/blob/9ad5d82f822b3cb67637f11be2e65c5662b66ec0/library/std/src/sys/unsupported/thread_local_dtor.rs), but I don't want to propose adding anything wasm-bindgen-specific to the standard library at this point. Can you be sure that the documentation for the thread support is updated with this caveat though? (and perhaps update the existing caveats to no longer indicate that stacks/tls allocatitons are leaked)

Finally after `__wbindgen_thread_destroy` is called can the instance's stack pointer get reset to 0 to prevent any accidental reuse?

Oh also for the-main-thread-may-block case I'm assuming that the majority of use cases have the browser main thread as the first thread anyway so I think it's ok to not gate this support behind a new env var.

alexcrichton commented:
Yeah I figure that since the glue does the allocation it may as well do the deallocation. Running the destructors is still an open question but that's more of a standard library and/or libc library integration issue than anything else. It's clear where we would execute them with a function like this, it's just not clear how we get access to them.

jrvidal commented:
I've tackled all of the above except for the extra tests, which I'll do next.

> Can you be sure that the documentation for the thread support is updated with this caveat though? (and perhaps update the existing caveats to no longer indicate that stacks/tls allocatitons are leaked)

Hm, is thread support documented anywhere? Do you mean the raytracer example in the [guide](https://github.com/rustwasm/wasm-bindgen/blob/main/guide/src/examples/raytrace.md)?

alexcrichton commented:
Yeah I'm imagining that this would update some of the wording around https://github.com/rustwasm/wasm-bindgen/blob/9b0d40c7a9330c7d36de2eb0c1fb4cf462cab86d/guide/src/examples/raytrace.md#caveats which mentions tls destructors and un-free-able stacks.

jrvidal commented:
OK, I've added some tests and updated the docs. The tests helped me catch a couple of bugs :sweat_smile: 

* Not calculating the base of the extra static data correctly, [[ref]](https://github.com/rustwasm/wasm-bindgen/pull/2769/commits/c9c8b2ff55a9f1c09ef3404897ba61373d00971c#diff-5c5b1e81729d4b11ddbb12a9e2cd96a0d5eee5bd25d4948363ca4fe157142467R251-R255)
* Not saving `__tls_base` for its subsequent deallocation. [[ref]](https://github.com/rustwasm/wasm-bindgen/pull/2769/commits/c9c8b2ff55a9f1c09ef3404897ba61373d00971c#diff-5c5b1e81729d4b11ddbb12a9e2cd96a0d5eee5bd25d4948363ca4fe157142467R354-R355)

Also I changed a couple of details:
* The temporary, shared stack is now aligned, which I supposed is better for performance? [[ref]](https://github.com/rustwasm/wasm-bindgen/pull/2769/commits/c9c8b2ff55a9f1c09ef3404897ba61373d00971c#diff-5c5b1e81729d4b11ddbb12a9e2cd96a0d5eee5bd25d4948363ca4fe157142467R146-R147)
* `tls.base` gets set to `i32::MIN` instead of 0, to make sure that we trigger memory access errors instead of overwriting valid memory. [[ref]](https://github.com/rustwasm/wasm-bindgen/pull/2769/commits/c9c8b2ff55a9f1c09ef3404897ba61373d00971c#diff-5c5b1e81729d4b11ddbb12a9e2cd96a0d5eee5bd25d4948363ca4fe157142467R399-R400)

What I'm going to do next is to run our internal test suite against this branch, as a form of "fuzzing".

alexcrichton commented on crates/threads-xform/tests/basic.wat:
```diff
@@ -0,0 +1,150 @@
+  (func $__wbindgen_thread_destroy (type 0)
+    global.get 1
+    i32.const -2147483648
+    i32.ne
+    if  ;; label = @1
+      global.get 1
+      i32.const 128
+      call $__wbindgen_free
+    else
+    end
```
This looks like it's trying to support multiple invocations of `__wbindgen_thread_destroy`, but I think we can probably skip that? In this destructor I think it's ok to assume it's the first and only time it's running (and basically have "UB" if it isn't)

alexcrichton commented on crates/threads-xform/tests/basic.wat:
```diff
@@ -0,0 +1,150 @@
+    i32.const -2147483648
+    global.set 1
+    global.get 3
+    if  ;; label = @1
```
Like above I think it's fine to skip this `if`

alexcrichton reviewed (commented):
Thanks for adding all this! It's definitely much easier to review seeing the output of the pass.

jrvidal commented on crates/threads-xform/tests/basic.wat:
```diff
@@ -0,0 +1,150 @@
+    else
+    end
+    i32.const 128
+    i32.const 4
+    drop
```
If I'm not mistaken, this corresponds to something that was already there: the fact that we ignore `__tls_align` [[ref]](https://github.com/rustwasm/wasm-bindgen/blob/main/crates/threads-xform/src/lib.rs#L311-L312). I think it shouldn't be too difficult to fix properly though.

alexcrichton replied:
Ah right I forgot about that! In that case no worries it's fine to leave in.

jrvidal commented:
Thanks for your prompt review, I think I've addressed all your comments, except for the one about `__tls_align`. I haven't had a chance to run our internal test suite yet. Not sure if this is something you want to block on, but I'll definitely try to do it tomorrow.

alexcrichton reviewed (approved):
Thanks again for this!

jrvidal commented:
:tada: Thanks for all your support.
